hotui_server/lib/hotui_server.dart

```diff
@@ -0,0 +1,256 @@
+  final String platformKernelDill =
+      options['platform'] ?? 'platform_strong.dill';
+  final CompilerOptions compilerOptions = CompilerOptions()
+    ..sdkRoot = sdkRoot
+    ..fileSystem = fileSystem
+    ..packagesFileUri =
+        convertFileOrUriArgumentToUri(fileSystem, options['packages'])
+    ..sdkSummary = sdkRoot.resolve(platformKernelDill)
+    ..verbose = options['verbose']
+    ..target = createFrontEndTarget(
+      options['target-model'],
+      trackWidgetCreation: true,
+    );
+  setVMEnvironmentDefines(<String, String>{}, compilerOptions);
+
+  // Create incremental compiler.
+  final Uri initializeFromDillUri =
+      Uri.file(options['initialize-from-dill'] ?? options['output-dill']);
+  final IncrementalCompiler incrementalCompiler = IncrementalCompiler(
+    compilerOptions,
+    entrypointUri,
+    initializeFromDillUri: initializeFromDillUri,
+  );
+
+  // Compile the initial component
+  final Component component =
+      await incrementalCompiler.compile(entryPoint: entrypointUri);
+  component.computeCanonicalNames();
+  incrementalCompiler.accept();
+
+  // Ready for hot UI.
+  HotUIService(
+    incrementalCompiler,
+    component,
+    File(initializeFromDillUri.toFilePath()).parent,
+  );
+}
+
+/// This service avoids re-parsing dart code by maintaining a single full
+/// component and performing updates to this component in place.
+///
+/// Hot ui works by specifiny a libraryId, classId, methodId, and new method
+/// body. The hot UI service uses this information to produce a partial kernel
+/// file containing the updated code. Notably, this is done without modifying
+/// source on disk using an inplace edit of the kernel AST.
+///
+/// Since the hot ui service may accept any libraryId, it must hold a full
+/// component in memory. Then, after modifying the AST, it manually trims the
+/// component to contain only the modified library.
+class HotUIService {
+  HotUIService(this.incrementalCompiler, this.component, this.outputDirectory) {
+    stdin
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen(onMessage);
+    print(kReadyMessage);
+  }
+
+  final IncrementalCompiler incrementalCompiler;
+  final Component component;
+  final Directory outputDirectory;
+
+  // Example invocation:
+  //
+  //    {"class":"_DemoItem","method":"build","library":"package:flutter_gallery/gallery/home.dart","methodBody":"Text('Hello, World')"}
+  Future<void> reloadMethod(String libraryId, String classId, String methodId,
+      String methodBody) async {
+    final Procedure procedure = await incrementalCompiler.compileExpression(
+        methodBody,
+        // This is not yet flexible for non-build expressions.
+        <String>['context'],
+        <String>['BuildContext'],
+        libraryId,
+        classId,
+        false);
+    component.transformChildren(
+        BodyReplacementTransformer(libraryId, classId, methodId, procedure));
```

Comment:
at a first glance this seems to be rather scary looking proposition - updating ast in such a drastic fashion outside of frontend. What is the overall problem being solved here? Who is going to be providing methodBody to this method? It might be faster than frontend to compile source(it would be nice to understand bottlenecks though - are they around file reading/serialization?), but does it provide as much safety and reliability than using frontend itself?

Reply:
thanks, I would include frontend/kernel folks to it as they might be interested in how their stuff is being(or will be) used :-)

Reply:
Sent out a document with the design and initial plan (looking for LGTM) - sorry for the delay on that, I should have done it first